tests/README.md

```diff
@@ -0,0 +1,1647 @@
+
+The commands bellow must be run from the ILIAS web root directory.
+
+<a name="execute-ilias-installation-bound-tests"></a>
+#### Execute ILIAS installation bound tests
+```
+./libs/composer/vendor/bin/phpunit -c ./Services/PHPUnit/config/PhpUnitConfig.xml \
+	--colors=always \
+	--no-globals-backup \
+	--report-useless-tests \
+	--disallow-todo-tests \
+	--group needsInstalledILIAS
+```
+
+<a name="execute-all-tests"></a>
+#### Execute all tests
+```
+./libs/composer/vendor/bin/phpunit -c ./Services/PHPUnit/config/PhpUnitConfig.xml \
+	--colors=always \
+	--report-useless-tests \
+	--disallow-todo-tests \
+	--no-globals-backup
+```
+
+<a name="execute-only-installation-unbound-tests"></a>
+#### Execute only installation unbound tests
+```
+./libs/composer/vendor/bin/phpunit -c ./Services/PHPUnit/config/PhpUnitConfig.xml \
+	--colors=always \
+	--no-globals-backup \
+	--report-useless-tests \
+	--disallow-todo-tests \
+	--exclude-group needsInstalledILIAS
+```
+
+<a name="guidelines"></a>
+## Guidelines
+
+<a name="foreword"></a>
+### Foreword
+Maybe the reader is asking him self why this guidelines refers to a book
+which writes about JUnit testing in Java 8.
+
+The reason for this is that unit testing in his very nature is the same in
+every language. The second reason is that no recently written books are
+on the market which describe modern unit testing with PHPUnit.
+
+There is no major reason why this particular book was taken. For the initial author
+this book seemed like a good starting point to create the unit test investigation.
+
+<a name="naming"></a>
+### Naming
+"Rework test names and code to tell stories." [1, Chap. 4]
+This means to treat the tests as a specification which tells everything
+about the behaviours of the unit under test.
+
+<a name="namespace"></a>
+#### Namespace
+The test class should always life in the same namespace as the test subject.
+For example the `ILIAS\HTTP\Cookies\CookieJarWrapperTest` 
+and the implementation `ILIAS\HTTP\Cookies\CookieJarWrapper` are in the same 
+namespace.
+
+<a name="class"></a>
+#### Class
+The filename of the test class should always be named like 
+*\<class name of the implementation\>Test.php*. Furthermore, the test class 
+should always be named as the class which is tested by the unit test class.
+For example the real class is called *Car* the corresponding test class would be 
+named *CarTest* and the filename *CarTest.php*.
+
+<a name="unit-test-method"></a>
+#### Unit-Test (Method)
+The method name must describe what your test is doing. For example the name 
+"testSomeBasics" is not really saying much about the test. It is also possible 
+that the test is actually testing multiple behaviours because of the 
+generic name.
+
+Some good, more descriptive names have the following forms:
+- "doingSomeOperationGeneratesSomeResult"       [1, Chap. 4]
+- "someResultOccursUnderSomeCondition"          [ebenda]
+- "whenDoingSomeBehaviourThenSomeResultOccurs"  [ebenda]
+
```

Comment:
I think an addition would help here: 
The method should always cover **only one** use case. Therefore multiple use cases should separated into different methods. See also [Good tests are FIRST](#good-tests-are-first)